### DISCUSSION TITLE
[Fast Data] Real Time Updater on error behaviour

### DISCUSSION THREAD

**Discussion by lucascanna** [mia-platform/community#342] [Q&A] [answered]
### Which section?

Console

### Which is the version?

11.7.2

### Description

Hi!

As far as I understand from documentation, the behaviour of the Real Time Updater in case it encounters an error while processing the Kafka message can be configured in two ways:
- if I set `PAUSE_TOPIC_CONSUMPTION_ON_ERROR ` to `true`: it will stop consuming from the topic where it encountered the error
- if I set `PAUSE_TOPIC_CONSUMPTION_ON_ERROR ` to `false`: it will crash

In both cases, if the problem is caused by a malformed message on the topic, the only way to solve it is to manually reset the offset of the Real Time Updater's consumer group.

Is it possible to let the RTU skip the message in case it's malformed (and put it in a Dead Letter Queue)?
I am aware that this would lead to issues related to the order the messages are processed but there are use cases where this is not an issue (e.g. If I only have inserts in my projections).

Thanks!

**Comment by danibix95** [comment, marked as answer]
Hi, right now Real-Time Updater plugin stops processing messages whenever an invalid one is received, since it is an unexpected situation.  
We specifically designed the service in this way because it is expected to process all the changes that occurs on the System of Record. Thus, we do not want the service to skip any change event, since they are all equally important for later generating Single Views.

Events produced on the ingestion topics are usually generated by Change Data Capture systems, which have their own specific format for all the messages they produce. We support some CDC [change event formats](https://docs.mia-platform.eu/docs/fast_data/inputs_and_outputs#ingestion-message) and it is also possible to provide a custom message adapter for formats not currently supported. 

Since Real-Time Updater reads event from Kafka, it is possible to use custom services in place of CDCs to write change events on ingestion topics. These implementation may generate invalid messages for whatever reason (e.g. missing data for composing the message).  
Consequently, knowing this details, I would recommend to implement an ingestion message **validation logic** on the custom producer, so that it is ensured that only valid messages are produced on the ingestion topics and the downstream components can process them.

**Reply by lucascanna** [reply]
Hi @danibix95,

thanks for the detailed answer!

I completely agree that in most of the cases discarding ingestion messages is not convinient and so, it is better to stop the consumption. 

However, there are use cases where discarding ingestion messages is acceptable and preferrable wrt stopping the entire flow. In this scenario, I may think to simply activate an alarm mechanism that notifies me about the error.

Take it as an idea of new feature: allow the user to choose the error behaviour.
